### ISSUE TITLE
NJS-103: unexpected message type 1 received at position 54 of packet 8372

### ISSUE THREAD

**Issue by damianhorex** [oracle/node-oracledb#1652]
1. What versions are you using?

    Oracle 19c Enterprise edition version 19.21.0.0.0

    process.platform: linux
    process.version: v20.11.0
    process.arch: x64
    require('oracledb').versionString: 6.4.0
    require('oracledb').oracleClientVersionString: undefined

3. Is it an error or a hang or a crash?

    Error

5. What error(s) or behavior you are seeing?

    NJS-103: unexpected message type 1 received at position 54 of packet 8372
        at Object.throwErr (/node_modules/oracledb/lib/errors.js:617:10)
        at ExecuteMessage.processMessage (/node_modules/oracledb/lib/thin/protocol/messages/base.js:216:14)
        at ExecuteMessage.processMessage (/node_modules/oracledb/lib/thin/protocol/messages/withData.js:91:13)
        at ExecuteMessage.process (/node_modules/oracledb/lib/thin/protocol/messages/base.js:195:12)
        at ExecuteMessage.decode (/node_modules/oracledb/lib/thin/protocol/messages/base.js:186:10)
        at Protocol._decodeMessage (/node_modules/oracledb/lib/thin/protocol/protocol.js:78:17)
        at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
        at async Protocol._processMessage (/node_modules/oracledb/lib/thin/protocol/protocol.js:153:9)
        at async ThinConnectionImpl._execute (/node_modules/oracledb/lib/thin/connection.js:194:7)
        at async ThinConnectionImpl.execute (/node_modules/oracledb/lib/thin/connection.js:907:14)

I am not entirely confident there is no user error playing a role. The original system was written using oracledb v1.12.2, and we are currently working on moving to v6.4. It is entirely possible I am missing something, but was hoping posting this might provide some insight. The NODE_ORACLEDB_DEBUG_PACKETS output is attached. Thank you.

[oracledb_debug_pkt_8372.txt](https://github.com/oracle/node-oracledb/files/14578539/oracledb_debug_pkt_8372.txt)

**Comment by sharadraju** [comment]
@damianhorex Thanks for using node-oracledb!

To debug this issue, please provide the following:

- A smaller reproducible test case
- Is this issue happening consistently?
- Have you tried this with a slightly earlier release like node-oracledb 6.2 or 6.3

This will help us in narrowing down the root cause.

**Comment by sudarshan12s** [comment]
Yes some information on minimal program,  especially the sql command issued which throws the error would help .  I think the database used is `Oracle 19c Enterprise edition version 19.21.0.0.0`

**Comment by damianhorex** [comment]
-     A smaller reproducible test case
I will work on producing a smaller test case but it may take some time. This issue has come up while testing a large amount of raw telemetry data and it is occurring deeper in the processing chain. I was hoping the dump provided would help. If I am reading things correctly, it seems that packet  8372 is a database response to an insert of image header data into our OCAMS_IMAGE_HEADER table? At the moment some inserts are succeeding, some are failing, so I am looking at isolating differences in the data itself. I currently do not have the specs to decode the response in packet 8372 which is one of the reasons I posted.
I did have a co-worker sanity check the download of the debug data, he mentioned it was not working for him yesterday, but did work this morning. Maybe try again? I will still attach again below.
-     Is this issue happening consistently?
Yes, I have consistently replicated it using 6.2, 6.3, and 6.4
-     Have you tried this with a slightly earlier release like node-oracledb 6.2 or 6.3
Yes, my latest tests have shown a similar problem


[oracledb_debug_pkt_8372.txt](https://github.com/oracle/node-oracledb/files/14590832/oracledb_debug_pkt_8372.txt)

Thanks!

**Comment by sudarshan12s** [comment]
@damianhorex , Thanks for sharing your findings. What i am seeing is that it looks you have a insert DML with large number of inbinds and 5 outbinds (NUMBER + 4 RAW) . While processing the response for outbind values, we are seeing some difference with what client decodes it as .. Correct me if its not what your test is trying to do..

We will meanwhile try to reproduce it locally . The database version is 19.21.0.0.0 right?

**Comment by sudarshan12s** [comment]
Can you share the following information if possible.
1. Does this test work in [thick](https://node-oracledb.readthedocs.io/en/latest/user_guide/initialization.html#enabling-node-oracledb-thick-mode-on-linux-and-related-platforms) mode?
2. The insert DML fails for first time?
3.  Do you change the input bind types for this insert DML subsequent execution or the bind information is provided with types fixed..

**Comment by damianhorex** [comment]
Thank you for your replies,
Yes, the database version is 19.21.0.0.0.
I will work on getting you answers to your other questions.

**Comment by damianhorex** [comment]
> @damianhorex , Thanks for sharing your findings. What i am seeing is that it looks you have a insert DML with large number of inbinds and 5 outbinds (NUMBER + 4 RAW) . While processing the response for outbind values, we are seeing some difference with what client decodes it as .. Correct me if its not what your test is trying to do..
> 
This is correct in this instance.

>     1. Does this test work in [thick](https://node-oracledb.readthedocs.io/en/latest/user_guide/initialization.html#enabling-node-oracledb-thick-mode-on-linux-and-related-platforms) mode?
> 

I just tested in thick mode and I no longer encountered the error.

Note: I am still working on putting together a small test case that replicates the issue but wanted to share the results of the test in thick mode.

**Comment by anthony-tuininga** [comment]
Can you try using a *differently* named bind variable for the returning clause? You are using the same name for input as well as the output for the returning clause -- and that isn't well supported. If I recall correctly, even though thick mode "works" it doesn't actually modify the output value. Can you confirm?

**Comment by anthony-tuininga** [comment]
I just confirmed it. You can't use bind variables with the same names before the returning clause and after the returning clause. In thick mode the value returned is not updated. We will ensure this situation doesn't raise that strange exception but in the meantime you can simply use different bind variable names before and after the RETURNING keyword to avoid this issue.

**Comment by damianhorex** [comment]
Ok, will do. Thank you for your time.

**Comment by manibhushan** [comment]
![image](https://github.com/oracle/node-oracledb/assets/3900481/6c162ee7-60a6-4455-bc8a-dc294a9a7434)
We are experiencing the same error code on our project. This happens while reading more rows (fetchMoreRows) on the resultSet. Is there any workaround available for this?

**Comment by sudarshan12s** [comment]
> ![image](https://private-user-images.githubusercontent.com/3900481/319721434-6c162ee7-60a6-4455-bc8a-dc294a9a7434.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTIyODgwOTYsIm5iZiI6MTcxMjI4Nzc5NiwicGF0aCI6Ii8zOTAwNDgxLzMxOTcyMTQzNC02YzE2MmVlNy02MGE2LTQ0NTUtYmM4YS1kYzI5NGE5YTc0MzQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDQwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA0MDVUMDMyOTU2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YmI5NjJjNDIxOWM0MjQ5YjQ5MGI4ZGUyMjA5YzJiYTIzNzIxNWVjOTAwN2FlOTFmZjAzNjk5YTg5OTgwZGNhMSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.i331mgPji-bdgHoMW4suKyNAJnHUHPTGBqrPA1_oLDM) We are experiencing the same error code on our project. This happens while reading more rows (fetchMoreRows) on the resultSet. Is there any workaround available for this?

Can you check if your sql has INOUT same bind variable name (below example has `str_val`) appearing in input and also in RETURNING INTO clause like this. If this is the case, you need to use different names..

` "INSERT INTO nodb_dmlreturn (id, name) VALUES (:int_val, :str_val) RETURNING id, name into :rid, :str_val"; binds = { int_val: 6, rid: { type: oracledb.NUMBER, dir: oracledb.BIND_OUT }, str_val: { type: oracledb.STRING, val: "A different test string", dir: oracledb.BIND_INOUT, maxSize: 200 } };`

**Comment by manibhushan** [comment]
```
SELECT
  PAYMENT_DATE , INVOICE_NO , PO_NUMBER , INVOICE_AMOUNT, DUE_DATE ,
  INVOICE_STATUS , PAYMENT_AMOUNT , CHECK_AMOUNT , CHECK_NO ,
  VENDOR_ID , INVOICE_DATE , PAYMENT_CURRENCY , SOURCE , CURRENCY ,
  DISCOUNT_DUE_DATE , RECEIVED_DATE, CARRIER_ID, UPDATE_DATE,
  SEGMENT_TYPE, TRANSPORT_MODE, LAST_INVOICE_OPERATIONS_REMARK
FROM
  ERPCUSTOBJS.ERP_OTM_SHP_INV_VW
WHERE
  (CARRIER_ID, 0) IN (('', 0), ('1001565', 0), ('1011907', 0))
AND ((SOURCE IN ('OTHER') AND (SEGMENT_TYPE IN ('HWY', 'DRC', 'MAT', 'DDR', 'LHL', 'ORC', 'CLA', 'SPT', 'EQP', 'ODR', 'ACR', 'NMV', 'MSC', 'XTN')))
OR (SOURCE IN ('OTHER-UYSN', 'OTHER-INS', 'OTHER-LGS')))
AND UPDATE_DATE >= TIMESTAMP '2024-04-01 00:03:57'
AND INVOICE_STATUS <> 'Voided'
```
It's not an INSERT statement that we are running. It's a SELECT query as above. The table has rows in excess of 50k and we are reading the rows in batches of 300 using the fetchMoreRows on resultset that's returned by this query. It doesn't fail during intitial batches but when it has read multiple thousand rows at some point it crashes throwing NJS-103 on fetchMoreRows.

**Comment by sudarshan12s** [comment]
> ```
> SELECT
>   PAYMENT_DATE , INVOICE_NO , PO_NUMBER , INVOICE_AMOUNT, DUE_DATE ,
>   INVOICE_STATUS , PAYMENT_AMOUNT , CHECK_AMOUNT , CHECK_NO ,
>   VENDOR_ID , INVOICE_DATE , PAYMENT_CURRENCY , SOURCE , CURRENCY ,
>   DISCOUNT_DUE_DATE , RECEIVED_DATE, CARRIER_ID, UPDATE_DATE,
>   SEGMENT_TYPE, TRANSPORT_MODE, LAST_INVOICE_OPERATIONS_REMARK
> FROM
>   ERPCUSTOBJS.ERP_OTM_SHP_INV_VW
> WHERE
>   (CARRIER_ID, 0) IN (('', 0), ('1001565', 0), ('1011907', 0))
> AND ((SOURCE IN ('OTHER') AND (SEGMENT_TYPE IN ('HWY', 'DRC', 'MAT', 'DDR', 'LHL', 'ORC', 'CLA', 'SPT', 'EQP', 'ODR', 'ACR', 'NMV', 'MSC', 'XTN')))
> OR (SOURCE IN ('OTHER-UYSN', 'OTHER-INS', 'OTHER-LGS')))
> AND UPDATE_DATE >= TIMESTAMP '2024-04-01 00:03:57'
> AND INVOICE_STATUS <> 'Voided'
> ```
> 
> It's not an INSERT statement that we are running. It's a SELECT query as above. The table has rows in excess of 50k and we are reading the rows in batches of 300 using the fetchMoreRows on resultset that's returned by this query. It doesn't fail during intitial batches but when it has read multiple thousand rows at some point it crashes throwing NJS-103 on fetchMoreRows.

Is this coming with 6.4 version? Can you upload the network traces dumped on console after setting this env, `NODE_ORACLEDB_DEBUG_PACKETS=1`

**Comment by manibhushan** [comment]
yes it's on 6.4 but it was also happening on lower versions which is essentially why we upgraded. it's difficult to get the network traces you are asking for because we only get this issue on prod env where we have a large number of rows being synced. on the testing env we haven't noted this happening. do you need the logs only when this issue happens or just in general.

**Comment by sudarshan12s** [comment]
> yes it's on 6.4 but it was also happening on lower versions which is essentially why we upgraded. it's difficult to get the network traces you are asking for because we only get this issue on prod env where we have a large number of rows being synced. on the testing env we haven't noted this happening. do you need the logs only when this issue happens or just in general.

Failed logs would help to get more data .    Can you confirm these for now:

- I see columns types are just simple Number, VARCHAR2, DATE. and no CLOB/BLOB seem to be involved.   
-  Any high level flow of application steps on DB operations would help so that we can try out too. 
- I think the size of 300 rows returned is not huge and its mostly few KB's as i see from column's ..
- Is this same scenario worked earlier in  [thick](https://node-oracledb.readthedocs.io/en/latest/user_guide/initialization.html#enabling-node-oracledb-thick-mode-on-linux-and-related-platforms) mode.

**Comment by manibhushan** [comment]
- All the column types are Number, VARCHAR2, DATE.
- We have two systems one of which uses SQL DB while the other uses Mongo DB. For our application requirements these two DBs need to be in sync. We have a job that runs and reads data from the SQL DB using the oracledb package and then populates in the MongoDB. The read op on SQL DB, like I specified earlier, happens in batches. There are no writes on the SQL DB in the entire process.
- The size of rows is not much, the table contains simple data.
- We haven't tried the thick mode yet. Testing that again would need deployment to our prod env and we can't risk deploying anything to prod without proper investigation as this is a critical process for our system.

**Comment by sudarshan12s** [comment]
> * All the column types are Number, VARCHAR2, DATE.
> * We have two systems one of which uses SQL DB while the other uses Mongo DB. For our application requirements these two DBs need to be in sync. We have a job that runs and reads data from the SQL DB using the oracledb package and then populates in the MongoDB. The read op on SQL DB, like I specified earlier, happens in batches. There are no writes on the SQL DB in the entire process.
> * The size of rows is not much, the table contains simple data.
> * We haven't tried the thick mode yet. Testing that again would need deployment to our prod env and we can't risk deploying anything to prod without proper investigation as this is a critical process for our system.

This issue is tracked separately [here](https://github.com/oracle/node-oracledb/issues/1661).

**Comment by sharadraju** [comment]
The fix for the original issue raised is available in node-oracledb v6.5.
